cmd/agent/app/reporter/client_metrics.go

```diff
@@ -0,0 +1,208 @@
+	lastUpdated           time.Time
+	batchSeqNo            int64
+	fullQueueDroppedSpans int64
+	tooLargeDroppedSpans  int64
+	failedToEmitSpans     int64
+}
+
+// ClientMetricsReporter is a decorator that emits data loss metrics on behalf of clients.
+type ClientMetricsReporter struct {
+	wrapped       Reporter
+	clientMetrics *clientMetrics
+	shutdown      chan struct{}
+
+	// map from client-uuid to *lastReceivedClientStats
+	lastReceivedClientStats sync.Map
+}
+
+// WrapWithClientMetrics creates ClientMetricsReporter.
+func WrapWithClientMetrics(reporter Reporter, mFactory metrics.Factory) *ClientMetricsReporter {
+	cm := new(clientMetrics)
+	metrics.MustInit(cm, mFactory.Namespace(metrics.NSOptions{Name: "client_stats"}), nil)
+	r := &ClientMetricsReporter{
+		wrapped:       reporter,
+		clientMetrics: cm,
+	}
+	go r.expireClientMetrics()
+	return r
+}
+
+// EmitZipkinBatch delegates to underlying Reporter.
+func (r *ClientMetricsReporter) EmitZipkinBatch(spans []*zipkincore.Span) error {
+	return r.wrapped.EmitZipkinBatch(spans)
+}
+
+// EmitBatch processes client data loss metrics and delegates to the underlying reporter.
+func (r *ClientMetricsReporter) EmitBatch(batch *jaeger.Batch) error {
+	r.updateClientMetrics(batch)
+	return r.wrapped.EmitBatch(batch)
+}
+
+// Close stops background gc goroutine for client stats map.
+func (r *ClientMetricsReporter) Close() {
+	close(r.shutdown)
+}
+
+func (r *ClientMetricsReporter) expireClientMetrics() {
+	const (
+		frequency = 15 * time.Minute
+		ttl       = time.Hour
+	)
+	ticker := time.NewTicker(frequency)
+	defer ticker.Stop()
+	for {
+		select {
+		case <-ticker.C:
+			t := time.Now()
+			var size int64
+			r.lastReceivedClientStats.Range(func(k, v interface{}) bool {
+				stats := v.(*lastReceivedClientStats)
+				stats.lock.Lock()
+				defer stats.lock.Unlock()
+
+				if !stats.lastUpdated.IsZero() && t.Sub(stats.lastUpdated) > ttl {
+					r.lastReceivedClientStats.Delete(k)
+				}
+				size += 1
+				return true // keep running through all values in the map
+			})
+			r.clientMetrics.ConnectedClients.Update(size)
+		case <-r.shutdown:
+			return
+		}
+	}
+}
+
+func (r *ClientMetricsReporter) updateClientMetrics(batch *jaeger.Batch) {
+	clientUUID := findClientUUID(batch)
+	if clientUUID == "" {
+		return
+	}
+	batchSeqNo := batch.SeqNo
+	if batchSeqNo == nil {
+		return
+	}
+	entry, found := r.lastReceivedClientStats.Load(clientUUID)
+	if !found {
+		entry, _ = r.lastReceivedClientStats.LoadOrStore(clientUUID, &lastReceivedClientStats{})
+	}
+	clientStats := entry.(*lastReceivedClientStats)
+	clientStats.update(*batchSeqNo, batch.Stats, r.clientMetrics)
+}
+
+func (s *lastReceivedClientStats) update(
+	batchSeqNo int64,
+	stats *jaeger.ClientStats,
+	metrics *clientMetrics,
+) {
+	s.lock.Lock()
+	defer s.lock.Unlock()
+
+	if s.batchSeqNo >= batchSeqNo && !wrapped(s.batchSeqNo, batchSeqNo) {
+		// ignore out of order batches, the metrics will be updated later
```

Comment:
The batch you are processing has a lower sequence number than a batch you've seen before, so, you aren't updating the metrics later, are you? I think you are just using the metrics from the highest known seq number instead.

Reply:
I will have unit tests to work through the edge cases.